workshop1/slides/src/slides.md

```diff
@@ -0,0 +1,664 @@
+* `a` is a *type variable* (could be any type)
+* `Nothing, Just` are *both* constructors
+  * `Nothing :: Maybe a`
+  * `Just :: a -> Maybe a`
+
+## The List Type
+
+* How do we represent a list of candidates?
+* Use singly-linked lists:
+  * `[]` is the empty list
+  * `:` lets you append values to the front
+  * `[a,b,c]` == `(a : b : c : [])`
+
+![](images/list.png){ height=40% }
+
+## The Data Model
+
+* Define our basic candidate type
+
+```haskell
+data Candidate = Candidate
+  { ...
+  } deriving Show
+```
+
+## The Data Model
+
+* Define our basic candidate type
+
+```haskell
+data Candidate = Candidate
+  { constituency :: String
+  , party        :: String
+  , sex          :: String
+  , averageAge   :: Double
+  } deriving Show
+```
+
+## The Data Model
+
+* Define our basic candidate type
+
+```haskell
+data Candidate = Candidate
+  { constituency :: String
+  , party        :: String
+  , sex          :: String
+  , averageAge   :: Double
+  } deriving Show
+```
+
+* Add a notion of failure
+
+## The Data Model
+
+* Define our basic candidate type
+
+```haskell
+data Candidate = Candidate
+  { constituency :: String
+  , party        :: String
+  , sex          :: String
+  , averageAge   :: Double
+  } deriving Show
+```
+
+* Add a notion of failure
+
+```haskell
+type DataModel = Maybe Candidate
+```
+
+## The Data Model
+
+* Define our basic candidate type
+
+```haskell
+data Candidate = Candidate
+  { constituency :: String
+  , party        :: String
+  , sex          :: String
+  , averageAge   :: Double
+  } deriving Show
+```
+
+* Add a notion of failure
+
+```haskell
+type DataModel = Maybe Candidate
+```
+
+* Create a list of candidates
+
+## The Data Model
+
+* Define our basic candidate type
+
+```haskell
+data Candidate = Candidate
+  { constituency :: String
+  , party        :: String
+  , sex          :: String
+  , averageAge   :: Double
+  } deriving Show
+```
+
+* Add a notion of failure
+
+```haskell
+type DataModel = Maybe Candidate
+```
+
+* Create a list of candidates
+
+```haskell
+type DataModel = [Maybe Candidate]
+```
+
+
+
+
+# Data Transformation
+
+## Goal
+
+* Write some functions to:
+  * Convert CSV into data model
+  * Convert data model into JSON
+* Demonstrate power of `Generic`
+* Introduce `map`
+
+### Approach
+
+* Write a function `parseLine :: CSVLine -> Maybe Candidate`
+* Write a function `parseFile :: CSV -> DataModel`
+* Generate a function `encode :: DataModel -> JSON`
+* Apply these to every row
+
+![](images/data-transformation.pdf)
+
+
+## Writing `parseLine :: CSVLine -> Maybe Candidate`
+
+```haskell
+type CSVLine = String -- type alias
+
+parseLine :: CSVLine -> Maybe Candidate
+parseLine line = makeCandidate (undefined)
+
+makeCandidate :: [String] -> Maybe Candidate
+makeCandidate fields = undefined
+
+fromMaybeAge
+  :: (Double -> Candidate)
+  -> Maybe Double
+  -> Maybe Candidate
+fromMaybeAge partialCandidate maybeAge = undefined
+```
+
+* `_` means "for every other value"
```

Comment:
This line doesn't fit here anymore.

Reply:
Move this line to the slide with `intToWord` implementation

Reply:
Done